### ISSUE TITLE
js: Consider adding Wasm support.

### ISSUE THREAD

**Issue by shurcooL** [gopherjs/gopherjs#833]
In https://github.com/hajimehoshi/gopherwasm, @hajimehoshi created an experimental wrapper that implements the `syscall/js` API using the `github.com/gopherjs/gopherjs/js` package in GopherJS, and using the `syscall/js` package itself in Wasm.

This gives me an idea for something that might be possible for the `github.com/gopherjs/gopherjs/js` package: to add an alternative implementation of the same API, but using the `syscall/js` package.

The GopherJS compiler would never use it, only the Go compiler would. It would help by making all packages that currently import `github.com/gopherjs/gopherjs/js` buildable for Wasm (with overhead, of course), until they can be updated to use `syscall/js` directly. This can be especially helpful for [browser API bindings](https://github.com/gopherjs/gopherjs/wiki/bindings) (e.g., see https://github.com/dominikh/go-js-dom/issues/57).

At this point this is just an idea. It needs investigation to see if it'd be viable. I wanted to create the issue so we can track it.

@hajimehoshi Have you already thought about this?

**Comment by hajimehoshi** [comment]
Hi, thank you for reaching for gopherwasm :-)

So your idea is to make a syscall/js wrapper with the GopherJS API? I have a mixed feeling. As libraries using GopherJS will (and should) move to Wasm in near or far future, I'm not sure we should make and keep such glue layer. Your suggestion would be useful for some existing libs with GopherJS, but how many such libraries there are?

**Comment by hajimehoshi** [comment]
This is the opposite suggestion to you, but how about making gopherwasm a member of GopherJS repositories? I'd like to, but I think this would conflicts with your suggestion.

**Comment by hajimehoshi** [comment]
Thank you, I'll move gopherwasm to a new repository under gopherjs org! Some questions:

* TBH, there should be a better name than gopherwasm, but I don't have a good idea. As the repository will be under `github.com/gohperjs` path, a shorter name would work but `wasm` would be too short. I'd be happy if you have a good suggestion.
* Probably the same license as GopherJS would be preferable. If you are not against, I'll change the license from Apache License 2.0 to BSD 2-clause.

**Comment by myitcv** [comment]
As I discussed with @bketelsen on Slack earlier, the one reservation I would have is that the semantics of the two APIs are quite different with respect to callbacks (sync vs async). The issue with a common API therefore is that it hides the fact these semantics are quite different.

**Comment by paralin** [comment]
@myitcv Go generally takes a synchronous approach, and you can abstract any of the async into sync APIs can you not? So, we could probably standardize around one API surface with some glue.

**Comment by myitcv** [comment]
@paralin 

> and you can abstract any of the async into sync APIs can you not?

Unless I'm missing something, once you've "gone async" you can't go back.

Consider an API function exposed to JS from Go includes a return value. If the scheduling of the function call is async, what return value can be expected?

**Comment by paralin** [comment]
@myitcv you can't expect one, that's why you abstract it. I can think of a way to write it with a channel. GopherJS would of course unwind the stack and return, and when the async value is emitted, the routine would be woken up and resumed.

That's of course probably slow but it's one example of an easy way to go async -> sync by leaning on the runtime.

**Comment by myitcv** [comment]
@paralin 

I'm unclear how your proposal works for a function call from within JS code. If that JS code is already written to expect a function that has a return value, what do you do? My use case is React, where as part of its API it expects a function that has a return value. Here's just one example from the API:

https://reactjs.org/docs/react-component.html#shouldcomponentupdate

In any case, there will be a change in the WASM approach during the Go 1.12 cycle:

https://github.com/golang/go/issues/26045

**Comment by flimzy** [comment]
[This](https://github.com/go-kivik/pouchdb/blob/master/bindings/pouchdb.go#L79-L95) is how I turn async JS code into sync Go code.

**Comment by theclapp** [comment]
> GopherJS would of course unwind the stack and return

Right.  It returns.  That's the end.  As far as the caller — in this case the browser itself firing an event — is concerned, it has returned.  The GopherJS scheduler can _call it again_ and it'll pick up where it left off, due to the magic of GopherJS, but as far as the browser's concerned, it returned earlier and that was that.

I mean, you're welcome to try it and prove me wrong (and seriously that'd be fine; maybe my code was wrong, or things have changed), but I'm pretty sure you'll find that what you're proposing won't work.

Also, in my experience "promises" count as returning.

Source: I was trying to write an event handler that opened a new window based on a mouse-click (or something along those lines; this was a while ago).  You _can_ do that, if you do it right then, in the immediate call-chain of the event handler.  If you try it at any other time (say, in a promise "then" or in some other goroutine), the browser says "Nope, you can only open a new window in response to a user action.  This call occurred elsewhere, therefore it's not a direct result of a user action, so you can't do it."

Again, if you can show code to prove us wrong, that'd be totally fine.  Just reporting my experience.

**Comment by paralin** [comment]
@theclapp @myitcv Okay, I think I understand the problem better now. You're calling something in JavaScript that expects an immediate value, but you can't do that because calling into Go requires asynchronous behavior. This makes sense, and agree that there's probably not a good solution to fix that directly.

**Comment by dmitshur** [comment]
I think by now it has become more clear that it's unlikely someone will want to spend the time on implementing this.

GopherJS now supports `syscall/js` API (PR #908), which means Go libraries can start implementing support for both Go WebAssembly and GopherJS simultaneously by using the `syscall/js` API. An example of this is the work on v2 of `honnef.co/go/js/dom` package (see https://github.com/dominikh/go-js-dom/issues/57#issuecomment-493640644).

The `syscall/js` API has a more future-friendly outlook, so that seems like a better direction to invest in, rather than trying to prolong the usefulness of the `github.com/gopherjs/gopherjs/js` API.

**Comment by hajimehoshi** [comment]
> The syscall/js API has a more future-friendly outlook, so that seems like a better direction to invest in, rather than trying to prolong the usefulness of the github.com/gopherjs/gopherjs/js API.

+1
